### PR TITLE
Dependency.get_variable method

### LINKED ISSUE #4957: RFC: Add a generic variable getter method for dependencies
We now have 4 different kinds of dependencies, cmake, pkg-config, config-tool, and declare_dependency. pkg-config and config-tool have a method that allow getting arbitrary variables, cmake might but I haven't looked closely. Some dependencies can be found multiple ways (cmake or config-tool for example). Because of that it would be nice to have a generic getter method, such as

```meson
dep = dependency('dep')
use_thing = dep.get_variable(cmake : 'CMakeVariable', pkg-config : 'pkg-config-variable', config-tool : 'config-tool-argument')
```
Which would be implemented such that each kind of dependency would respect the proper keyword.

@jpakkane, opinions?

### PR DESCRIPTION
**Pull request by dcbaker**
This provides a new method to the Dependency class that simplifies dependencies that can be found multiple ways, and provides a way to get arbitrary variables from cmake based dependencies. Each dependency type as a keyword argument that can be passed for it's type.

As a follow up we may want to deprecated the `get_*type*_variable` methods in favor of this generic one.

Fixes #4957

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -439,13 +440,21 @@ def partial_dependency_method(self, args, kwargs):
...
+        return self.held_object.get_variable(**kwargs)
+
+
 class InternalDependencyHolder(InterpreterObject, ObjectHolder):
     def __init__(self, dep, pv):
         InterpreterObject.__init__(self)
         ObjectHolder.__init__(self, dep, pv)
         self.methods.update({'found': self.found_method,
-                             'version': self.version_method,
+                             'get_varaible': self.variable_method,
```
Comment (jon-turney) [inline mesonbuild/interpreter.py]: varaible -> variable

File: docs/markdown/Dependencies.md [hunk 2]
```diff
@@ -64,6 +64,26 @@ pkg-config files. Meson has autodetection support for some of these,
 and they are described [later in this
 page](#dependencies-with-custom-lookup-functionality).
 
+# Arbitrary variables from dependencies that can be found multiple ways
+
+*Note* new in 0.51.0
+
+When you need to get an arbarary variables from a dependency that can be
+found multiple ways, and you don't want to constraint the type you can use
```
Comment (jon-turney) [inline docs/markdown/Dependencies.md]: constraint -> constrain

File: docs/markdown/Dependencies.md [hunk 3]
```diff
@@ -64,6 +64,26 @@ pkg-config files. Meson has autodetection support for some of these,
...
+and config-tool based variables.
+
+```meson
+foo_dep = dependency('foo')
+var = foo.get_variable(cmake : 'CMAKE_VAR', pkgconfig : 'pkg-config-var', configtool : 'get-var', default : 'default')
+```
+
+It accepts the keywords 'cmake', 'pkgconfig', 'pkgconfig_define',
+'configtool', and 'default'. If the value cannot be found by the
+implementaiton specific method and default is defined then that value will be
```
Comment (jon-turney) [inline docs/markdown/Dependencies.md]: implementaiton -> implementation

File: docs/markdown/Reference-manual.md [hunk 4]
```diff
@@ -2173,6 +2173,14 @@ an external dependency with the following methods:
    - includes: any include_directories
    - sources: any compiled or static sources the dependency has
 
+ - `get_variable(cmake : str, pkgconfig : str, configtool : str, default :
+   str, pkgconfig_define : [str, str]) *(Added in 0.51.0)* A generic
+   variable getter method, which repalces the get_*type*_variable methods.
+   This allows one to get the variable from a dependency without konwing
```
Comment (jon-turney) [inline docs/markdown/Reference-manual.md]: konwing -> knowing

File: docs/markdown/Dependencies.md [hunk 5]
```diff
@@ -64,6 +64,26 @@ pkg-config files. Meson has autodetection support for some of these,
...
+returned, otherwise an error raised. pkgconfig_define has the same meaning
+as 'define_variable' to `get_pkgconfig_variable`.
+
```
Comment (jon-turney) [inline docs/markdown/Dependencies.md]: This should probably also specify what happens when the variable name is not specified for the dependency type which was found (e.g. found by cmake, but cmake: kwarg is absent).

File: docs/markdown/Dependencies.md (hunk 2 above)
Comment (jon-turney) [reply docs/markdown/Dependencies.md]: arbarary -> arbitrary

File: docs/markdown/Dependencies.md (hunk 5 above)
Comment (dcbaker) [reply docs/markdown/Dependencies.md]: I tried to resolve this, are you happy with the rewording?

File: docs/markdown/Dependencies.md [hunk 6]
```diff
@@ -64,6 +64,30 @@ pkg-config files. Meson has autodetection support for some of these,
 and they are described [later in this
 page](#dependencies-with-custom-lookup-functionality).
 
+# Arbitrary variables from dependencies that can be found multiple ways
+
+*Note* new in 0.51.0
+
+When you need to get an arbitrary variables from a dependency that can be
+found multiple ways, and you don't want to constrain the type you can use
```
Comment (jon-turney) [inline docs/markdown/Dependencies.md]: Does this need a comma after 'constrain the type' to mark a new clause?  Should this sentence better be in passive voice?

File: docs/markdown/Dependencies.md [hunk 7]
```diff
@@ -64,6 +64,30 @@ pkg-config files. Meson has autodetection support for some of these,
...
+
+```meson
+foo_dep = dependency('foo')
+var = foo.get_variable(cmake : 'CMAKE_VAR', pkgconfig : 'pkg-config-var', configtool : 'get-var', default : 'default')
+```
+
+It accepts the keywords 'cmake', 'pkgconfig', 'pkgconfig_define',
+'configtool', and 'default'. 'pkgconfig_define' works just like the
+'define_variable' argument to `get_pkgconfig_variable`. When this
+method is invoked the keyword coresponding to the underlying type
```
Comment (jon-turney) [inline docs/markdown/Dependencies.md]: coresponding -> corresponding
the first underlying in 'underlying type of the underlying dependency' is superfluous?

File: docs/markdown/Dependencies.md (hunk 5 above)
Comment (jon-turney) [reply docs/markdown/Dependencies.md]: Makes sense

File: docs/markdown/snippets/dependency_get_variable_method.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
+## Dependency objects now have a get_variable method
+
+This is a generic replacement for type specific variable getters such as
+`ConfigToolDependency.get_configtool_variable` and
+`PkgConfigDependency.get_pkgconfig_variable`, and is the only way to query
+such variables from cmake dependencies.
+
+This method allows you to get variables without knowing the kind of
+dependnecy you have.
```
Comment (jon-turney) [inline docs/markdown/snippets/dependency_get_variable_method.md]: dependnecy -> dependency

File: docs/markdown/snippets/dependency_get_variable_method.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+```meson
+dep = dependency('could_be_cmake_or_pkgconfig')
+# cmake returns 'YES', pkg-config returns 'ON'
+if ['YES', 'ON'].contains(dep.get_variable(pkg-config : 'var-name', cmake : 'COP_VAR_NAME', default : 'NO'))
+  error('Cannot build your project when dep is build with var-name support')
```
Comment (jon-turney) [inline docs/markdown/snippets/dependency_get_variable_method.md]: build -> built?

File: mesonbuild/dependencies/base.py [hunk 10]
```diff
@@ -518,6 +525,23 @@ def get_configtool_variable(self, variable_name):
...
         return self.type_name
 
+    def get_variable(self, *, cmake: typing.Optional[str] = None, pkgconfig: typing.Optional[str] = None,
+                     configtool: typing.Optional[str] = None, default: typing.Optional[str] = None,
+                     pkgconfig_define: typing.Optional[typing.List[str]] = None) -> typing.Union[str, typing.List[str]]:
+        if configtool:
+            try:
+                # in the not required case this will be '' (empty string) if
+                # the variable is not found, in that case we want to continue
+                # and not return ''
```
Comment (jon-turney) [inline mesonbuild/dependencies/base.py]: This seems a bit unfortunate as it would preclude the value from being an empty string in this case?

**Comment by jon-turney** [comment]
pleeze trun spleelhcck on :)

**Comment by dcbaker** [comment]
:/ I tried, i'm just getting an super helpful error along the lines of "spell checking: error"

File: mesonbuild/dependencies/base.py (hunk 10 above)
Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: You're right, I've changed it to force an error in that case so we can differentiate the two.

**Comment by dcbaker** [comment]
Hopefully that addresses everything.

**Comment by jpakkane** [comment]
When I first saw this I thought the `default` was for the default variable name to look up rather than the default value that is returned. So if you would do something like this

```meson
v = dep.get_variable(default: 'name', cmake: 'customname')
```

to a pkg-config dependency it would look up the pkg-config variable named `name` rather than return the value `name`.

Is it just me? Does this seem like a reasonable confusion to make? If yes, maybe it should could be called `default_value` instead?

**Comment by jpakkane** [comment]
And expanding on that, should there be a way to specify a default key? Something like:

```meson
x = dep.get_variable('name')
```

One would imagine most variables would have the same names. They should at least in the case of pkgconfig and internal deps if we add the possibility to set variables to internal deps.

**Comment by dcbaker** [comment]
I guess when I thought about it pkg-config and cmake names never are equal.

**Comment by jpakkane** [comment]
If we change `default` to `default_value` then we should be able to merge this. Making the first posar into a default value (or adding a `default_name`) later can be done without breaking backwards compatibility, right?

**Comment by dcbaker** [comment]
sure, I'll change `default` to `default_value`. I think we can add a posarg later if we decide that would be useful, and that shouldn't break the API since it currently takes 0 posargs.

**Pull request merged**